File: examples/ruby/covenant-vault/CovenantVault.runar.rb
require 'runar'

class CovenantVault < Runar::SmartContract
  prop :owner, PubKey
  prop :recipient, Addr
  prop :min_amount, Bigint

  def initialize(owner, recipient, min_amount)
    super(owner, recipient, min_amount)
    @owner = owner
    @recipient = recipient
    @min_amount = min_amount
  end

  runar_public sig: Sig, amount: Bigint, tx_preimage: SigHashPreimage
  def spend(sig, amount, tx_preimage)
    assert check_sig(sig, @owner)
    assert check_preimage(tx_preimage)
    assert amount >= @min_amount
  end
end

File: examples/zig/covenant-vault/CovenantVault.runar.zig
const runar = @import("runar");

pub const CovenantVault = struct {
    pub const Contract = runar.SmartContract;

    owner: runar.PubKey,
    recipient: runar.Addr,
    minAmount: i64,

    pub fn init(owner: runar.PubKey, recipient: runar.Addr, minAmount: i64) CovenantVault {
        return .{
            .owner = owner,
            .recipient = recipient,
            .minAmount = minAmount,
        };
    }

    pub fn spend(self: *const CovenantVault, sig: runar.Sig, txPreimage: runar.SigHashPreimage) void {
        runar.assert(runar.checkSig(sig, self.owner));
        runar.assert(runar.checkPreimage(txPreimage));

        const expectedOutput = runar.buildChangeOutput(self.recipient, self.minAmount);

        runar.assert(runar.bytesEq(runar.hash256(expectedOutput), runar.extractOutputHash(txPreimage)));
    }
};
